@here/harp-mapview/lib/text/Placement.ts

```diff
@@ -313,46 +345,118 @@ export function placePointLabel(
 
     if (label.bounds === undefined) {
         label.bounds = new THREE.Box2();
-        tmpMeasurementParams.outputCharacterBounds = undefined;
-        tmpMeasurementParams.path = undefined;
-        tmpMeasurementParams.pathOverflow = false;
-        tmpMeasurementParams.letterCaseArray = label.glyphCaseArray!;
-        textCanvas.measureText(label.glyphs!, label.bounds, tmpMeasurementParams);
     }
 
-    screenPosition.add(computePointTextOffset(label, scale, tmpTextOffset));
-    outScreenPosition.set(screenPosition.x, screenPosition.y, labelState.renderDistance);
-
-    // TODO: Make the margin configurable
-    tmpBox.copy(label.bounds!).expandByVector(pointLabelMargin);
-    tmpBox.min.multiplyScalar(scale);
-    tmpBox.max.multiplyScalar(scale);
-    tmpBox.translate(screenPosition);
-    tmp2DBox.set(
-        tmpBox.min.x,
-        tmpBox.min.y,
-        tmpBox.max.x - tmpBox.min.x,
-        tmpBox.max.y - tmpBox.min.y
-    );
-
-    // Check the text visibility.
-    if (!screenCollisions.isVisible(tmp2DBox)) {
-        return PlacementResult.Invisible;
-    }
+    // Setup measurements parameters for textCanvas.measureText().
+    tmpMeasurementParams.outputCharacterBounds = undefined;
+    tmpMeasurementParams.path = undefined;
+    tmpMeasurementParams.pathOverflow = false;
+    tmpMeasurementParams.letterCaseArray = label.glyphCaseArray!;
+
+    // NOTE:
+    // We may save some CPU time by firstly checking visibility of all possible text placements,
+    // their bounding boxes merged altogether, this may save time for further calculations.
+    // Would be enough to include TL and BR placements.
+
+    // Start with currently set alignment settings.
+    const basePlacement = {
+        h: label.layoutStyle!.horizontalAlignment,
+        v: label.layoutStyle!.verticalAlignment
+    };
+    // Placements options will be read from label.layoutStyle.placements in final solution.
+    const placements = possibleAnchorPlacementsCW;
+    // Find first anchor placement.
+    const idx = placements.findIndex(val => {
+        return val.h === basePlacement.h && val.v === basePlacement.v;
+    });
+    // If not listed in placement options list (idx === -1) will start from current alignment setup.
+    const placementsNum = placements.length;
+    let textOffset: THREE.Vector2;
+    // Will be true if all text placements are invisible.
+    let allInvisible: boolean = true;
+    let i = 0;
+    for (i = 0; i < placementsNum; ++i) {
+        const anchorPlacement = idx + i < 0 ? basePlacement : placements[(idx + i) % placementsNum];
+        // Override label text alignment for measurements and leave it so if passed collisions test.
```

Comment:
Better not touching the layout style, just pass the placement alternative to the computePointTextOffset function.

Reply:
See my comments above. I assume that last anchor placement (alignment) is stored in label object and this one in turn is persistent until label is removed or styling is changed (via theme rules - zoom level, etc.).

Reply:
You're losing here the preferred alignment coming from the style, overwriting it with the next succesful alignment option  that was found. 
This means next time you find that label (e.g. imagine you pan away from the label till it's not visible, rotate 90deg and pan back) you'll take that alignment and try to apply it first, even if the the original placement indicated by the style would be successful in this case.
To do this cleanly we should save the current alignment for persistent labels. I say the best place is in TextElementState. Even though we're saving lots of stuff in TextElement, i don't think it's the right place to do it.

Reply:
> You're losing here the preferred alignment coming from the style, overwriting it with the next succesful alignment option that was found.

That is intentional! Imagine different situation, when label is rejected for while by the more important label (usual when rotating the camera at some tilt) then it comes back after a sec, we should then keep layout consistent and reduce text jumping as much as possible. Another situation is when label goes out of screen but very close to it, when it returns back, should it be totally new layout or rather recently viewed one?